src/resolve/resolved-dns-rr.c

```diff
@@ -826,6 +827,115 @@ static char *format_txt(DnsTxtItem *first) {
+        int r;
+
+        switch (i->key) {
+        case DNS_SVC_PARAM_KEY_ALPN: {
+                size_t sz = 0, offset = 0;
+                _cleanup_strv_free_ char **values_strv = NULL;
+                while (offset < i->length) {
+                        char *alpn;
+                        sz = (uint8_t) i->value[offset++];
+                        alpn = strndup((char *)&i->value[offset], sz);
+                        if (!alpn)
+                                return NULL;
+                        strv_push(&values_strv, alpn);
+                        offset += sz;
+                }
+                value = strv_join(values_strv, ",");
+                break;
+        }
+        case DNS_SVC_PARAM_KEY_PORT: {
+                uint16_t port = unaligned_read_be16(i->value);
+                r = asprintf(&value, "%" PRIu16, port);
+                if (r < 0)
+                        return NULL;
+                break;
+        }
+        case DNS_SVC_PARAM_KEY_IPV4HINT: {
+                const struct in_addr *addrs = (struct in_addr*) &i->value;
+                _cleanup_strv_free_ char **values_strv = NULL;
+                for (size_t n = 0; n < i->length / sizeof (struct in_addr); n++) {
+                        char *addr;
+                        r = in_addr_to_string(AF_INET, (const union in_addr_union*) &addrs[n], &addr);
+                        if (r < 0)
+                                return NULL;
+                        strv_push(&values_strv, addr);
+                }
+                value = strv_join(values_strv, ",");
+                if (!value)
+                        return NULL;
+                break;
+        }
+        case DNS_SVC_PARAM_KEY_IPV6HINT: {
+                const struct in6_addr *addrs = (struct in6_addr*) &i->value;
+                _cleanup_strv_free_ char **values_strv = NULL;
+                for (size_t n = 0; n < i->length / sizeof (struct in6_addr); n++) {
+                        char *addr;
+                        r = in_addr_to_string(AF_INET6, (const union in_addr_union*) &addrs[n], &addr);
+                        if (r < 0)
+                                return NULL;
+                        strv_push(&values_strv, addr);
+                }
+                value = strv_join(values_strv, ",");
+                if (!value)
+                        return NULL;
+                break;
+        }
+        default: {
+                char *p;
+                p = value = new(char, 4 * i->length + 3);
+                *(p++) = '"';
+                for (size_t j = 0; j < i->length; j++) {
+                        if (i->value[j] < ' ' || IN_SET(i->value[j], ',', '\\', '"') ||
+                            i->value[j] >= 127) {
+                                *(p++) = '\\';
+                                *(p++) = '0' + (i->value[j] / 100);
+                                *(p++) = '0' + ((i->value[j] / 10) % 10);
+                                *(p++) = '0' + (i->value[j] % 10);
```

Comment:
just use cescape() or the related calls?

Reply:
This is intended to reproduce the dec-octet encoding described in appendix A to the rfc. It's basically copy-pasta'd from the TXT handling above. I think I'll leave it if that's alright?

Reply:
hmm, if this is already used at another place, then please add a helper call for it, and reference the RFC the algorithm is described in, in a comment.